drivers/pcf857x/pcf857x.c

```diff
@@ -0,0 +1,470 @@
+{
+    /* some parameter sanity checks */
+    assert(dev != NULL);
+    assert(params != NULL);
+    assert(params->exp < PCF857X_EXP_MAX);
+#if IS_USED(MODULE_PCF857X_IRQ)
+    assert(gpio_is_valid(params->int_pin));
+#endif
+
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init device data structure */
+    dev->params = *params;
+
+    switch (params->exp) {
+#if IS_USED(MODULE_PCF8574)
+        /**< PCF8574 8 bit I/O expander used */
+        case PCF857X_EXP_PCF8574: dev->pin_num = PCF8574_GPIO_PIN_NUM;
+                                  dev->params.addr += PCF8574_BASE_ADDR;
+                                  break;
+#endif
+#if IS_USED(MODULE_PCF8574A)
+        /**< PCF8574A 8 bit I/O expander */
+        case PCF857X_EXP_PCF8574A: dev->pin_num = PCF8574A_GPIO_PIN_NUM;
+                                   dev->params.addr += PCF8574A_BASE_ADDR;
+                                   break;
+#endif
+#if IS_USED(MODULE_PCF8575)
+        /**< PCF8575 16 bit I/O expander */
+        case PCF857X_EXP_PCF8575: dev->pin_num = PCF8575_GPIO_PIN_NUM;
+                                  dev->params.addr += PCF8575_BASE_ADDR;
+                                  break;
+#endif
+        default: return -PCF857X_ERROR_INV_EXP;
+    }
+
+#if IS_USED(MODULE_PCF857X_IRQ)
+    /* initialize the IRQ event object used for delaying interrupts */
+    dev->irq_event.event.handler = _irq_handler;
+    dev->irq_event.dev = dev;
+
+    for (unsigned i = 0; i < dev->pin_num; i++) {
+        dev->isr[i].cb = NULL;
+        dev->isr[i].arg = NULL;
+        dev->enabled[i] = false;
+    }
+
+    /* initialize the interrupt pin */
+    if (gpio_init_int(dev->params.int_pin,
+                      GPIO_IN_PU, GPIO_FALLING, _irq_isr, (void*)dev)) {
+        return -PCF857X_ERROR_INT_PIN;
+
+    }
+#endif /* MODULE_PCF857X_IRQ */
+
+    int res = PCF857X_OK;
+
+    _acquire(dev);
+
+    /* write 1 to all pins to switch them to INPUTS pulled up to HIGH */
+    dev->out = ~0;
+    res |= _write(dev, dev->out);
+
+    /* initial read all pins */
+    res |= _read(dev, &dev->in);
+
+    /* set all pin modes to INPUT and set internal output data to 1 (HIGH) */
+    dev->modes = ~0;
+
+    _release(dev);
+
+    return PCF857X_OK;
+}
+
+int pcf857x_gpio_init(pcf857x_t *dev, gpio_t pin, gpio_mode_t mode)
+{
+    /* some parameter sanity checks */
+    assert(dev != NULL);
+    assert(pin < dev->pin_num);
+
+    DEBUG_DEV("pin=%u mode=%u", dev, pin, mode);
+
+    switch (mode) {
+        case GPIO_IN_PD: DEBUG_DEV("gpio mode GPIO_IN_PD not supported", dev, mode);
+                         return -PCF857X_ERROR_INV_MODE;
+        case GPIO_OUT:   dev->modes &= ~(1 << pin); /* set mode bit to 0 */
+                         dev->out   &= ~(1 << pin); /* set output bit to 0 */
+                         break;
+        default:         dev->modes |= (1 << pin); /* set mode bit to 1 */
+                         dev->out   |= (1 << pin); /* set output bit to 1 */
```

Comment:
Now this also catches `GPIO_IN_PU` and `GPIO_OD`/`GPIO_OD_PU`, is this correct?

Reply:
Ups, simultaneous communication.

This I/O expander is a very simple device that doesn't have a direction register. In fact it physically supports only the GPIO modes `GPIO_IN_PU` and `GPIO_OD_PU`:
- Writing a 1 to an expander pin configures the pin as an input, which is pulled up to HIGH by a weak pull-up. When reading the pin, its value then depends on the actual voltage.
- Writing a 0 to an expander pin configures the pin as an output and actively drives the pin to LOW.

All other modes are only considered logically identical and behave only like `GPIO_IN`, `GPIO_OD` and `GPIO_OUT`.

Reply:
It's probably worth adding a comment about that

Reply:
> Now this also catches GPIO_IN_PU and GPIO_OD/GPIO_OD_PU, is this correct?

This affects only the initial level of the pin. If `GPIO_OUT` is used a 0 is written to the pin which configures the pin as physical output an drives it active LOW. In all other cases a 1 is written to the pin which configures it as physical input an the pin is pulled-up to HIGH by the weak pull-up. It would also be possible to use HIGH as initial level for all modes.

Reply:
> It's probably worth adding a comment about that - feel free to squash directly

It is documented here https://github.com/RIOT-OS/RIOT/blob/0b9b8c28e5bb9a16ca0b4d88e9ec27dded1b3514/drivers/include/pcf857x.h#L67. But I can add a comment in source code too.

Reply:
Comment added to source code in commit https://github.com/RIOT-OS/RIOT/pull/10430/commits/d02a87fe0ac10115b8edb977abcb323d5a8be5ac.